### DISCUSSION TITLE
ENV Vars not working opa as golang lib

### DISCUSSION THREAD

**Discussion by sied95** [open-policy-agent/community#221] [OPA and Rego]
Hi all, sorry if this is a silly question.

I was trying to use `opa.runtime().env.FOO` in a policy, using embedded opa inside golang as a lib, but that doesn't seem to work at all.
Is there something I'm missing here?

When I run `opa run` as the standard binary, works just fine.

Can confirm the envs are properly set.

````rego
verify_action(id) = action { 
  base_url_env := opa.runtime().env.MANAGEMENT_API_URL
  timeout := opa.runtime().env.MANAGEMENT_API_URL_TIMEOUT
  base_url := sprintf(base_url_env,[id])
  response := http.send({"method": "get", "url": base_url, "timeout": timeout})
  response.status_code == 200
  action := response.body.Action
} else = action {
 action := "Alert"
}
````

this only works when replacing the env per literal string.

**Comment by anderseknert** [comment]
Hi @Siedlarczyk ! [This](https://github.com/open-policy-agent/opa/issues/3548#issuecomment-859204146) should be what you're looking for.

**Reply by sied95** [reply]
Hi @anderseknert, that helps a bit, but not sure I'm on the right path.

As per the code I see that it requires *ast.Term

```// Construct a Rego object that can be prepared or evaluated.
	MGMT_API := ast.StringTerm(os.Getenv("API"))
	r := rego.New(
		rego.Query("FOO"),
		rego.Load([]string{policyPath}, nil),
		rego.Runtime(MGMT_API),
	)
```
and my policy is 

```
verify_action(id) = action { 
  base_url_env := opa.runtime().env.MGMT_API
  #base_url := sprintf(base_url_env,[id])
  response := http.send({"method": "get", "url": base_url_env})
  response.status_code == 200
  action := response.body.Action
} else = action {
    action := "Alert"
}
```
should this be enough?

**Reply by anderseknert** [reply]
The comment below in the same thread demonstrates how to do it: https://github.com/open-policy-agent/opa/issues/3548#issuecomment-859300078

You're inserting a String now, but you'll rather want to build an object of name -> value mappings, and put that mapping in an "env" key in the object you insert into the runtime.

**Reply by sied95** [reply]
Is it possible to refer to the sole string I've inserted?

**Reply by anderseknert** [reply]
Sure! The example iterates over all env vars.. you could just change that to use a single one:

```go
	// add env variables to rego query
	obj := ast.NewObject()
	env := ast.NewObject()
	for _, s := range os.Environ() {
		parts := strings.SplitN(s, "=", 2)
		if parts[0] == "MGMT_API" {
			env.Insert(ast.StringTerm(parts[0]), ast.StringTerm(parts[1]))
		}
	}
	ast := obj.Insert(ast.StringTerm("env"), ast.NewTerm(env))
```
I did not test this code, but it should give you an idea :)

**Reply by sied95** [reply]
ast := obj.Insert(ast.StringTerm("env"), ast.NewTerm(env))

I might be missing something, but obj.Insert returns nothing
<img width="875" alt="Captura de Tela 2022-06-21 às 10 31 11" src="https://user-images.githubusercontent.com/26888429/174811662-804b2033-39e4-4ee2-9188-6543bbf91f0b.png">

**Reply by sied95** [reply]
and the point would be like, considering that rego.Runtime expects an *ast.Term, and I can build one from a string, would it be possible to refer to that sole string in the runtime?
Something like
```
opa.runtime().MGMT_API
```

**Reply by anderseknert** [reply]
No, the `env` attribute is a key-value object representing the environment variables set in the environment. 

```
opa.runtime().env.MGMT_API
```
would be the way to do it.
